src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -506,19 +512,41 @@ private function validateAndDenormalize(string $currentClass, string $attribute,
             // To circumvent this behavior, integers are converted to floats when denormalizing JSON based formats and when
             // a float is expected.
             if (Type::BUILTIN_TYPE_FLOAT === $builtinType && \is_int($data) && false !== strpos($format, JsonEncoder::FORMAT)) {
-                return (float) $data;
+                return $this->denormalizationSuccess((float) $data);
             }
 
             if (('is_'.$builtinType)($data)) {
-                return $data;
+                return $this->denormalizationSuccess($data);
             }
         }
 
         if ($context[self::DISABLE_TYPE_ENFORCEMENT] ?? $this->defaultContext[self::DISABLE_TYPE_ENFORCEMENT] ?? false) {
-            return $data;
+            return $this->denormalizationSuccess($data);
+        }
+
+        $message = sprintf('The type of the "%s" attribute for class "%s" must be one of "%s" ("%s" given).', $attribute, $currentClass, implode('", "', array_keys($expectedTypes)), get_debug_type($data));
+
+        return $this->denormalizationFailure($attribute, $message, $context);
+    }
+
+    private function denormalizationSuccess($denormalizedValue)
+    {
+        if ($context[self::COLLECT_INVARIANT_VIOLATIONS] ?? false) {
+            return DenormalizationResult::success($denormalizedValue);
+        }
+
+        return $denormalizedValue;
+    }
+
+    private function denormalizationFailure($normalizedValue, string $message, array $context)
+    {
+        if ($context[self::COLLECT_INVARIANT_VIOLATIONS] ?? false) {
+            $violation = new InvariantViolation($normalizedValue, $message);
+
+            return DenormalizationResult::failure([$violation]);
```

Comment:
There is a bit of a mix-up here:
* the `$normalizedValue` parameter will actually receive the `$attribute`
* when instantiating the `DenormalizationResult` it should be of type `array<string, array<InvariantViolation>>`
```suggestion
    private function denormalizationFailure($normalizedValue, string $attribute, string $message, array $context)
    {
        if ($context[self::COLLECT_INVARIANT_VIOLATIONS] ?? false) {
            $violation = new InvariantViolation($normalizedValue, $message);

            return DenormalizationResult::failure([$attribute => [$violation]]);
```

And update the calls to be like:
```php
$this->denormalizationFailure($data, $attribute, $message, $context);
```

Reply:
Fixed.